pkg/ddl/ddl_worker.go

```diff
@@ -492,28 +484,150 @@ func (d *ddl) addBatchDDLJobs(tasks []*limitJobTask) error {
 
-		jobTasks = append(jobTasks, job)
+		jobs = append(jobs, job)
 		injectModifyJobArgFailPoint(job)
-		if !job.LocalMode {
-			d.initJobDoneCh(job.ID)
-		}
 	}
 
 	se.GetSessionVars().SetDiskFullOpt(kvrpcpb.DiskFullOpt_AllowedOnAlmostFull)
-
-	if tasks[0].job.LocalMode {
+	ddlSe := sess.NewSession(se)
+	localMode := tasks[0].job.LocalMode
+	if localMode {
+		if err = fillJobIDs(ctx, ddlSe, tasks); err != nil {
+			return err
+		}
 		for _, task := range tasks {
 			d.localJobCh <- task
 		}
 		return nil
 	}
-	return errors.Trace(insertDDLJobs2Table(sess.NewSession(se), true, jobTasks...))
+
+	if err = GenIDAndInsertJobsWithRetry(ctx, ddlSe, jobs); err != nil {
+		return errors.Trace(err)
+	}
+	for _, job := range jobs {
+		d.initJobDoneCh(job.ID)
+	}
+
+	return nil
+}
+
+// GenIDAndInsertJobsWithRetry generate job ID and inserts DDL jobs to the DDL job
+// table with retry. job id allocation and job insertion are in the same transaction,
+// as we want to make sure DDL jobs are inserted in id order, then we can query from
+// a min job ID when scheduling DDL jobs to mitigate https://github.com/pingcap/tidb/issues/52905.
+// so this function has side effect, it will set the job id of 'tasks'.
+func GenIDAndInsertJobsWithRetry(ctx context.Context, ddlSe *sess.Session, jobs []*model.Job) error {
+	return genIDAndCallWithRetry(ctx, ddlSe, len(jobs), func(ids []int64) error {
+		for idx := range jobs {
+			jobs[idx].ID = ids[idx]
+		}
+		return insertDDLJobs2Table(ctx, ddlSe, jobs...)
+	})
+}
+
+func fillJobIDs(ctx context.Context, ddlSe *sess.Session, tasks []*limitJobTask) error {
+	var allocatedIDs []int64
+	if err := genIDAndCallWithRetry(ctx, ddlSe, len(tasks), func(ids []int64) error {
+		allocatedIDs = ids
+		return nil
+	}); err != nil {
+		return errors.Trace(err)
+	}
+
+	for i, task := range tasks {
+		task.job.ID = allocatedIDs[i]
+	}
+	return nil
+}
+
+// genIDAndCallWithRetry generates global IDs and calls the function with retry.
+// generate ID and call function runs in the same transaction.
+func genIDAndCallWithRetry(ctx context.Context, ddlSe *sess.Session, count int, fn func(ids []int64) error) error {
+	var resErr error
+	for i := uint(0); i < kv.MaxRetryCnt; i++ {
+		resErr = func() (err error) {
+			if err := ddlSe.Begin(ctx); err != nil {
+				return errors.Trace(err)
+			}
+			defer func() {
+				if err != nil {
+					ddlSe.Rollback()
+				}
+			}()
+			txn, err := ddlSe.Txn()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			txn.SetOption(kv.Pessimistic, true)
+			forUpdateTS, err := lockGlobalIDKey(ctx, ddlSe, txn)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			txn.GetSnapshot().SetOption(kv.SnapshotTS, forUpdateTS)
+
+			m := meta.NewMeta(txn)
+			ids, err := m.GenGlobalIDs(count)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if err = fn(ids); err != nil {
+				return errors.Trace(err)
+			}
+			return ddlSe.Commit(ctx)
+		}()
+
+		if resErr != nil && kv.IsTxnRetryableError(resErr) {
```

Comment:
not sure what's the meaning of `ErrLockExpire`. I just want to cover the case that pessimistic lock is cleaned by other readers of GlobalIDKey. (If the only reader is the new DDL owner, we don't need to retry because it means current node should retire)

Reply:
see https://github.com/tikv/client-go/blob/d73cc1ed6503925dfc7226e8d5677ceb4c2fd6f1/txnkv/transaction/2pc.go#L1227-L1230, and handled in tidb with below, seems no special handling https://github.com/pingcap/tidb/blob/06e0e17d98abbe3f88f82cee7746c0940384364f/pkg/session/session.go#L1029-L1040

i guess the reason is: T1 lock it so no one can change it before forUpdateTS, after the lock is expired & it hasn't commit, if there is T2 lock it later, T1 commit > T2 forUpdateTS  > T1 forUpdateTS, one of them will report `write conflict` on commit

Reply:
I found this type of error `ERROR 1105 (HY000): tikv aborts txn: Error(Txn(Error(Mvcc(Error(PessimisticLockNotFound` which likely standing for the case that pessimistic lock is cleaned by other transactions. But I don't know if it's handled by `kv.IsTxnRetryableError`. Transaction is very complex 😂

Reply:
> But I don't know if it's handled by kv.IsTxnRetryableError

https://github.com/pingcap/tidb/blob/9044acbff3a189b26b52ffd2990ce7b4f59021e0/pkg/kv/error.go#L83-L85

only ErrTxnRetryable except write conflict

> Transaction is very complex 😂

yes, that's why i ask @lcwangchao @cfzjywxk to review this part

Reply:
default `MaxTxnTTL:             60 * 60 * 1000, // 1hour`, seems ok to let user retry it themself in this case, even DML transaction will abort

Reply:
I see `ManagedLockTTL` in client-go is 20s. Does it means if pessimistic lock left in TiKV while the node crashes, other node must wait at least 20s to clean up the lock? If lock owner crashes, submitting DDL will be paused for 20s, it's not friendly.

Reply:
seems we cannot avoid this on crash
- create table t(id int primary key, v int); insert into t values(1,1);
- mysql -uroot -h 127.0.0.1 -P4000 test -e "begin; update t set v=2 where id=1; select sleep(100);";
```
mysql -uroot -h 127.0.0.1 -P4000 test -e "select now(); begin; update t set v=2 where id=1; select sleep(100);";
+---------------------+
| now()               |
+---------------------+
| 2024-07-15 21:59:56 |
+---------------------+
ERROR 2013 (HY000) at line 1: Lost connection to MySQL server during query
```
- kill -9 tidb, immediately after previous step
- restart
- run immediately after resart 
```
mysql -uroot -h 127.0.0.1 -P4000 test -e "begin; update t set v=3 where id=1; commit; select now();"
+---------------------+
| now()               |
+---------------------+
| 2024-07-15 22:00:17 |
+---------------------+
```